### PR TITLE
multiple modifications

### PR DESCRIPTION
**Pull request by CY-dev**
Hi,

I have made several changes based on what we need in our use case. Please tell me if you have any questions, and merge them if you find them useful.

4 API changes to train method of Trainer and its subclasses:

- also added support for multi-output models by passing a list of output column names to label_col argument

- as a follow-up with mult-output support, also added loss_weights as an argument

- accept a keras optimizer instance to be passed to the "optimizer" argument
  (from keras.optimizers import Optimizer, serialize, deserialize to deal with the object internally)

- In the notebook environment, it seems parameter_server is not quite cleaned up when the trainer.train method gets interrupted, even with try-except approach described in issue #25. It still prints the "Address already in use" message. Therefore I added master_port as an argument to "train" method of Trainer classes as a work-around

2 internal implementation changes:

- for num_epoch > 1, instead of data itself make copies of iterator via itertools.tee function. This change could help avoid time-consuming "unionAll" method due to data copying

- simply coalesce instead of repartition when num_partitions = num_workers, since coalescing does not shuffle the data and require much less time than repartitioning.

### CODE REVIEW

File: distkeras/workers.py [hunk 1]
```diff
@@ -110,20 +129,24 @@ def start_prefetching_thread(self, iterator):
...
-                    feature_iterators = iterator_copies[:-1]
-                    label_iterator = iterator_copies[-1]
-                    X = [np.asarray([x[self.features_column[i]] for x in iterator]) 
-                        for i, iterator in enumerate(feature_iterators)]
-                    Y = np.asarray([x[self.label_column] for x in label_iterator])
-                    self.mini_batches.put([X, Y])
-        except Exception as e:
-            print(e)
-            self.is_prefetching = False
+        partition_iterators_all_epochs = tee(self.iterator, self.num_epoch)
```
Comment (JoeriHermans) [inline distkeras/workers.py]: Looks solid. However, was the wait induces by the unionAll really that high? I mean, Spark is just constructing DAG nodes right?

**Comment by JoeriHermans** [comment]
Hi,

Looks very solid, thanks. However I have some doubts with handling the epochs at a worker level as I did this before in a initial prototype. Is the unionAll significantly impairing the training procedure?

Also, thanks for fixing the docs :)

Joeri

**Comment by JoeriHermans** [comment]
Any updates? I see that a lot of commits have been pushed in the meantime.

**Comment by CY-dev** [comment]
Hi Joeri,

The unionAll is making copies of the data. So when initially the Spark Dataframe has 10 partitions, it will have 50 if we want to train 5 epochs using unionAll, which seems inefficient.

Thanks,
Congrui

**Comment by mcmathja** [comment]
@JoeriHermans The main issue is that calling `cache` on a dataframe forces Spark to materialize its current state in memory before continuing. Because the call to `unionAll` comes before that, you end up actually copying out the entirety of the duplicated data. That overhead gets quite bad when running a large number of epochs.

Would love to see this branch merged. 👍

**Comment by JoeriHermans** [comment]
Hi @mcmathja That makes sense, I only realize it now :) Thanks. I'm going to review the final code now and merge it when I have no further issues.

File: distkeras/workers.py [hunk 2]
```diff
@@ -262,13 +287,19 @@ def train(self, worker_id, iterator):
         self.pull()
         self.model.set_weights(self.center_variable)
         try:
+            sys.stderr.write("Debug: starting optimize...\n")
```
Comment (JoeriHermans) [inline distkeras/workers.py]: Could this debugging information be removed?

**Comment by JoeriHermans** [comment]
Looks very good guys, thanks a lot! Could you check the final comment I've made on the debug-messages? Also did the GPU allocation work?

File: distkeras/workers.py (hunk 2 above)
Comment (CY-dev) [reply distkeras/workers.py]: Just commented the debugging message

**Comment by JoeriHermans** [comment]
Merging. Very good job guys!

Joeri
